mypy/checker.py

```diff
@@ -2266,8 +2266,14 @@ def visit_decorator(self, e: Decorator) -> None:
                                                    [nodes.ARG_POS], e,
                                                    callable_name=fullname)
         sig = cast(FunctionLike, sig)
-        sig = set_callable_name(sig, e.func)
-        e.var.type = sig
+        if e.var.type is not None:
+            # We have a declared type, check it.
+            self.check_subtype(sig, e.var.type, e,
```

Comment:
@JukkaL Shouldn't this be other way around?
I thought that the declared type is expected to be more narrow/precise than the inferred.

Reply:
I think that this is right -- this is not a cast but a declaration. But it would be good to have a test case where the ordering of this check is important.

Reply:
The problem with this is that it will be then impossible to give a more precise type for a decorated function when it is known. From the discussion on python/peps tracker I somehow understood that this should work like a cast. I however could imagine situations where one might want to declare narrower argument types, so maybe this check is not needed and we can allow both ways?

Reply:
This can be used to declare a more precise type with respect to `Any` types (`Any` is considered a subtype of everything).

Reply:
OK, I see, I was thinking about hijacking this decorator for https://github.com/python/mypy/issues/2087